### DISCUSSION TITLE
Unity Package Manager support

### DISCUSSION THREAD

**Discussion by Dasparion** [KyryloKuzyk/PrimeTween#9] [FAQ] [answered]
Would be great to add package manager support for this plugin. Using package manager allows for cleaner updates and workflows.

**Comment by KyryloKuzyk** [comment]
PrimeTween already installs via Package Manager! Although, Asset Store has its caveats, so the library installs with the help of tgz archive. This lets me take the best from both words: it’s easy to find the demo scene because it installs into the Assets folder, and the library itself installs into the Packages with the full **source code** available. 

There are also other reasons that I described here: https://forum.unity.com/threads/primetween-high-performance-animations-and-sequences.1479609/#post-9231630

**Comment by KyryloKuzyk** [comment]
@SudoCat I totally understand the 'thumbs down' reaction. However, PrimeTween is not an open-source product in the classical sense of this word. It's an Asset Store product that is currently available for free, but may become paid in the future. And I don't want to give false expectation to my users by distributing PrimeTween as a scoped registry package. It would be dishonest to revoke access in the future for users who installed via UMP's scroped registry.

I understand that the current distribution method of Asset Store is not perfect, and I would love to see the UPM support added to Asset Store in the future. But for now, I'm trying to make the installation as seamless as possible with the tools available to the Asset Store publisher.

**Reply by SudoCat** [reply]
I understand and respect that choice for you as the creator, you can distribute it however you fancy. I mean no offense with the reaction, I just came across this issue while hoping for the same feature, and felt it important to add to community sentiment. I think you'll find a lot of folks, especially as time progresses, hoping for this option. 

Asset store is pain, package managers are nice yadayada. 

However on your point about changing to paid in the future: surely if you distribute on asset store, then switch to a payment model, you'll see similar issues? I see no reason you'd realistically need to even revoke access in that situation - you'd just stop publishing updates to that registry. Heck, Unity has done this to us enough times, we're used to it. I'll admit though, I have no experience of publishing on the asset store, I've only used package managers for distributing OSS.

If you are interested in considering the viability of a UPM distribution model, then you could always start with using OpenUPM while you're still in this early stage, and then if you wanted to make a premium one and maintain the UPM distribution, you could use your own [scoped registry with authentication](https://docs.unity3d.com/Manual/upm-config-scoped.html)

I understand this won't be a priority, but I do think you should give it further consideration, especially if you want to compete with the old guards of Tweening (which I'd be really keen to see! your API is much nicer than any others I've looked at)

**Reply by KyryloKuzyk** [reply]
By 'revoking access' I didn't mean to revoke access from all current users, which would be a nightmare. What I meant is to 'stop providing access to new users, while the old ones can continue using the library further'. With Asset Store, all previous users who got the product for free in the past will have access to all future updates, which is fair, and this is how I see the transition to the premium model.

But OpenUPM doesn't give an easy option to limit access to the users who have downloaded the product before. And the option of stopping publishing OpenUPM updates feels dishonest toward loyal customers.

I will give the UPM distribution method another thought and maybe I'll come up with a solution that will make everyone happy.

**Reply by bdovaz** [reply]
Any updates?

I like that in theory it is a package although it is not distributed as such and the clear performance improvement but it makes me quite backward to discard DOTween for this asset because you may want to change it to a payment model... 

DOTween also has its "Pro" asset but it is for more advanced cases that are not scripting but editor scripts to visually create paths, etc... Apart from integration with other assets I imagine through extension methods. But the "simple" use through scripting is kept free: https://github.com/Demigiant/dotween/blob/develop/LICENSE

Thanks!

**EDIT:** Just so it doesn't sound like I'm selfish not to value the effort of others, if you look at my profile you'll see that I'm an active contributor in several projects so I know what it's like to dedicate time to it without anything in return.

In fact, if it were open source I would surely help you with PRs.

**Reply by KyryloKuzyk** [reply]
@bdovaz 
> you may want to change it to a payment model

If this will happen (which is very unlikely), you'll still 'own' the product on Asset Store and will receive all future updates. I also like the idea of a PRO version more, and most likely the current PrimeTween version will stay free forever. I just don't want to make irreversible decisions for now. Releasing PrimeTween on OpenUPM or GitHub is one of those irreversible decisions.

My view on PrimeTween is that it's not an open-source project in the classical sense. It's an Asset Store product that comes with full source code and has a 'free' price tag. The same as with DOTween, you can use PrimeTween in free or [commercial projects](https://github.com/KyryloKuzyk/PrimeTween/discussions/27?converting=1) and you can modify the source code for your own use.

I very appreciate your desire to help with PRs! However, the hardest part for me with PrimeTween is not adding features, but deciding which features NOT to add to keep the API as simple and consistent as possible. I have my own specific vision for the project, which some people don't share (like the absence of [extension methods](https://github.com/KyryloKuzyk/PrimeTween/discussions/3)), but I like how things is turning up for now. So if you have a particular feature request, I would love to hear from you :)

**Reply by Romaleks360** [reply]
DOTween requirement to install it in the Assets folder and not via Package Manager is the number 1 reason I'm looking for other solutions. Libraries like these should be installable as a package with a single git link, period. It's just too much of a headache to use other distrubution methods. Right now I will choose LitMotion purely for this reason.

**Reply by KyryloKuzyk** [reply]
Yes, I agree that installation via the git link is the best option for the user. PrimeTween requires importing it from the Asset Store, but after that, the installation is seamless and invisible (it installs to the Packages folder). So I hope you'll have a chance to try PrimeTween :)

I'll think about releasing PrimeTween to OpenUPM in the future. But currently, it would require one extra step to my release pipeline.

Also, PrimeTween offers many more features and quality of life improvements in comparison to LitMotioin:
https://forum.unity.com/threads/primetween-high-performance-animations-and-sequences.1479609/page-6#post-9754969

**Reply by bdovaz** [reply]
@KyryloKuzyk do you still think the same? I'm thinking of replacing DOTween with PrimeTween but if you still disregard the fact that it is distributed by OpenUPM I don't see it as an alternative.

Thanks.

**Reply by KyryloKuzyk** [reply]
I don't have plans to make PrimeTween paid, so OpenUPM seems like a good option and I don't mind distributing the plugin with it.

The only thing that stops me now is an easy way to do automatic releases there. I would like to keep my repo private because I'm developing the Pro version in the same repo. Do you by chance know if there is a way to distribute a *.unitypackage or *.tgz file to UMP? Or can I just point OpenUMP to this file so it can turn it into a package automatically?
https://github.com/KyryloKuzyk/PrimeTween/blob/main/Benchmarks/Packages/com.kyrylokuzyk.primetween.tgz

> I don't see it as an alternative

DOTween also doesn't support OpenUMP, so why not?
You'll lose nothing and get a modern tween library in return :)

**Reply by arimger** [reply]
> I don't have plans to make PrimeTween paid, so OpenUPM seems like a good option and I don't mind distributing the plugin with it.
> 
> The only thing that stops me now is an easy way to do automatic releases there. I would like to keep my repo private because I'm developing the Pro version in the same repo. Do you by chance know if there is a way to distribute a *.unitypackage or *.tgz file to UMP? Or can I just point OpenUMP to this file so it can turn it into a package automatically? https://github.com/KyryloKuzyk/PrimeTween/blob/main/Benchmarks/Packages/com.kyrylokuzyk.primetween.tgz
> 
> > I don't see it as an alternative
> 
> DOTween also doesn't support OpenUMP, so why not? You'll lose nothing and get a modern tween library in return :)

Hi @KyryloKuzyk, I see potential solutions:
- create a public repository that is synchronized with your private one through GitHub Actions
   - this public repository can be used by OpenUPM to publish all releases
   - I'm not sure about the synchronizing tags, making releases, etc. but maybe it's worth exploring it
- create your own public Scoped Registry (using [Verdaccio](https://verdaccio.org/) or [Cloudsmith](https://cloudsmith.com/))
   - you can manually or through GitHub Actions or any other automation publish releases without making the repository public
   - you can also do it manually
   - you can automate splitting (using `git subtree split`) your repo to make a dedicated branch for publishing the package
   - your package will be outside OpenUPM but still can be used as a UPM dependency which is convenient
   - I'm currently using Cloudsmith for my personal needs and it works pretty well

**Comment by KyryloKuzyk** [comment, marked as answer]
@Dasparion @bdovaz @SudoCat @arimger 
Good news! PrimeTween is now available on NPM!

To install the plugin via Unity Package Manager, open the `Packages/manifest.json` file and add the `scopedRegistries` section. If you already have it, then just add a new `com.kyrylokuzyk` scope.
```
"scopedRegistries": [
  {
    "name": "npm",
    "url": "https://registry.npmjs.org/",
    "scopes": [
      "com.kyrylokuzyk"
    ]
  }
]
```

Then add this line to the `dependencies` section:
```
    "com.kyrylokuzyk.primetween": "1.3.0",
```
